### ISSUE TITLE
Can't skip NullResultGuardian to Wrap MockHandlerImpl

### ISSUE THREAD

**Issue by klevis** [mockito/mockito#350]
Returning NULL Java Class Primitives like Long is consider a desired behavior in my case. But I cannot get Null Object Primitives because of NullResultGuardian which converts Object Primitives Null to primitives every time. For ex Long=>null is converted to 0. 
Is it possible to skip this NullResultGuardian?
EX usage:
Mockito.doAnswer(new Answer<Long>() {
    Longanswer(InvocationOnMock invocation) {
       if(some cond){
       return null;
       }else return value.
    }
}).when(object).doSoemthing(someParameterValue);
Below is were actually the place where this is wrap is initialized. 
public class MockHandlerFactory {

    public InternalMockHandler create(MockCreationSettings settings) {
        InternalMockHandler handler = new MockHandlerImpl(settings);
        InternalMockHandler nullResultGuardian = new NullResultGuardian(handler);

        return new InvocationNotifierHandler(nullResultGuardian, settings);
    }
}

**Comment by ChristianSchwarz** [comment]
Can you provide a test case that includes the method signature of the stubbed methods? 

I'am not sure how your usecase looks like, and what doesn't work. Do you want to return `null` when your method with return type `Long` is invoked? 

Maybe you are looking for this -> `when(object.doSomething(arg)).thenReturn((Long)null)`

**Comment by klevis** [comment]
EX usage:
Mockito.doAnswer(new Answer() {
Longanswer(InvocationOnMock invocation) {
if(some condition is met ){
return null;
}else return notNullValue.
}
}).when(object).doSoemthing(someParameterValue);

Case you suggested always returns null which is not what I am looking for, because the return value(null or not) is decided at runtime(read from xls).

**Comment by ChristianSchwarz** [comment]
So you want to return `null` if a condition it satisfied, and otherwise a non null value should be returned? 

Maybe you want something like this:
```
@Test
public void test(){
  boolean condition = true/false;
  Long value;
  if (condition)
     value= null;
  else
     value=nonNullValue;
  when(obj.doSomething(param)).returnValue(value);
}
```

**Comment by klevis** [comment]
Suggested solution will not work because:
1-Logic for determined the value must be inside the mock method because that method will be invoke several times and each time the result must different
2-Besiede that will be called several time , each time the passed parameter to method is different , and condition is based on method parameter. So condition checking must be encapsulated inside the mock method.

`Mockito.doAnswer(new Answer() {
Long answer(InvocationOnMock invocation) {
String methodArg=(String)invocation.getArguments()[0];
if(methodArg.equals("1")){
return null;
}else return notNullValue.
}
}).when(object).doSoemthing(someParameterValue);`

**Comment by ChristianSchwarz** [comment]
If your answer implementation works for you, everthing is fine with mockito right?

**Comment by klevis** [comment]
No actually when I do what I wrote before I cannot get a null as answer because NullResultGuardian will not allow that. It return 0 for primitives. What we need is make use of this NullResultGuardian optional is config. I can help with pull request on this but first I need some insight.

**Comment by ChristianSchwarz** [comment]
I don't know if i follow, here is my last try: Can you rewrite the following test so it fails? I can't reproduce your described misbehaviour on a `null` return value.

```
interface Function {
	Integer apply(String arg);
}

@Test
public void test() throws Exception {
	Function f = mock(Function.class);
	when(f.apply(any())).then(returnLengthOrNull());

	assertThat(f.apply(null), is(nullValue()));
	assertThat(f.apply(""), is(0));
	assertThat(f.apply("x"), is(1));
}

private Answer<Integer> returnLengthOrNull() {
	return new Answer<Integer>() {
		@Override
		public Integer answer(InvocationOnMock i) throws Throwable {
			String text = i.getArgumentAt(0, null);
			if (text == null)  //or any other condition
				return null;
			return text.length();
		};
	};
}
```

**Comment by klevis** [comment]
You are right everything is fine with mockito was just a misunderstanding on my part.

Follow method in NullResultGuardian:
 public Object handle(Invocation invocation) throws Throwable {
        Object result = this.delegate.handle(invocation);
        Class returnType = invocation.getMethod().getReturnType();
        return result == null && returnType.isPrimitive()?(new HandyReturnValues()).returnFor(returnType):result;
    }
it is checking return type of real method than in my case was primitive like **long and not Long**(was changed last time I checked :))
that's why I couldn't see null. NullResultGuardian is returning default values for non wrapper primitive types as consequence doing the right thing.

Sorry for inconvenience
